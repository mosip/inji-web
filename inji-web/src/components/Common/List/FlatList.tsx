import React, {Fragment} from "react";

type FlatListProps<T> = {
    data: T[];
    renderItem: (item: T) => React.ReactNode;
    keyExtractor: (item: T) => string;
    testId: string;
    onEmpty?: React.ReactNode;
};

export function FlatList<T>({
                                data,
                                renderItem,
                                keyExtractor,
                                testId,
                                onEmpty
                            }: Readonly<FlatListProps<T>>) {
    const gridStyle = `grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 w-full gap-1`;

    if (data.length === 0) {
        if (onEmpty === undefined)
            return <div data-testid={"empty-list"}>No items available</div>
        return (
            <Fragment>
                {onEmpty}
            </Fragment>
        )
    }

    return (
        <div className={gridStyle} data-testid={testId}>
            {data.map((item) => (
                <Fragment key={keyExtractor(item)}>
                    {renderItem(item)}
                </Fragment>
            ))}
        </div>
    );
}
