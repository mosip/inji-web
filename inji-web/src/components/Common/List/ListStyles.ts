const ListStyles = {
    gridLayout: "grid sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 w-full gap-1"
}

export default ListStyles