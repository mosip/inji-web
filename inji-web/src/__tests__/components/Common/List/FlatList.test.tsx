import React from "react";
import {render, screen} from "@testing-library/react";
import {FlatList} from "../../../../components/Common/List/FlatList";

describe("FlatList Component", () => {
    type Item = {
        id: string;
        name: string;
    }
    const sampleData: Item[] = [
        {id: "1", name: "Item 1"},
        {id: "2", name: "Item 2"},
        {id: "3", name: "Item 3"}
    ];

    const keyExtractor = (item: Item) => item.id;
    const renderItem = (item: Item) => <div>{item.name}</div>;

    it('should match the snapshot', () => {
        const {asFragment} = render(
            <FlatList
                data={sampleData}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                testId="flatlist"
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders all items", () => {
        render(
            <FlatList
                data={sampleData}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                testId="flatlist"
            />
        );
        sampleData.forEach(({name}) => {
            expect(screen.getByText(name)).toBeInTheDocument();
        });
    });

    it("renders fallback message if no items", () => {
        render(
            <FlatList
                data={[]}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                testId="flatlist"
            />
        );
        expect(screen.getByText("No items available")).toBeInTheDocument();
        expect(screen.getByTestId("empty-list")).toHaveTextContent("No items available");
    });

    it("renders custom empty state if provided", () => {
        render(
            <FlatList
                data={[]}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                testId="flatlist"
                onEmpty={<div data-testid="custom-empty">Nothing here</div>}
            />
        );
        expect(screen.getByTestId("custom-empty")).toBeInTheDocument();
        expect(screen.getByText("Nothing here")).toBeInTheDocument();
    });

    it("applies correct responsive grid classes", () => {
        render(
            <FlatList
                data={sampleData}
                renderItem={renderItem}
                keyExtractor={keyExtractor}
                testId="flatlist"
            />
        );
        const list = screen.getByTestId("flatlist");
        // Check if grid class contains responsive col definitions
        expect(list.className).toMatch(/grid-cols-1/);
        expect(list.className).toMatch(/sm:grid-cols-2/);
        expect(list.className).toMatch(/md:grid-cols-2/); // 2/3 of 3 = 2
        expect(list.className).toMatch(/lg:grid-cols-3/);
    });
});
