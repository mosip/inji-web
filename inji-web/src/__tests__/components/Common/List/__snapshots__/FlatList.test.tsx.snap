// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatList Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-2 lg:grid-cols-3 w-full gap-1"
    data-testid="flatlist"
  >
    <div>
      Item 1
    </div>
    <div>
      Item 2
    </div>
    <div>
      Item 3
    </div>
  </div>
</DocumentFragment>
`;
